src/app.test.js

```diff
@@ -1,15 +1,52 @@
-// import { createApp } from "./app";
-// import userevent from '@testing-library/user-event'
-
-test("renders the todo app", () => {
-  //   // setup
-  //   const root = ...
-  //   const container = createApp(root)
-  //   // act
-  //   const input = queryByPlaceholderText('Please enter todo here')
-  //   userevent.type(input, 'new task')
-  //   const addButton = queryByText('Add')
-  //   userevent('click', addButton)
-  //   // assert
-  expect(true).tobee(true);
+import { screen, getByText } from "@testing-library/dom";
+import userEvent from "@testing-library/user-event";
+import createApp from "./app";
+// import userEvent from "@testing-library/user-event";
+
+test("renders empty todo list when starts", () => {
+  // setup
+  const root = document.body;
+  createApp(root);
+  const todoList = screen.getByTestId("todo-list");
+  // act
+  // assert
+  expect(todoList).toBeEmptyDOMElement();
+});
+
+test("shows new item in todo list when added", () => {
+  // setup
+  const input = screen.getByPlaceholderText("Please enter todo here");
+  const addButton = screen.getByTestId("add-button");
+  const inputText = "Take wife for a walk!";
+
+  // act
+  userEvent.type(input, inputText);
+  userEvent.click(addButton);
+
+  // assert
+  const todoList = screen.getByTestId("todo-list");
+  const todoTextElement = getByText(todoList, "Take wife for a walk!");
+  expect(todoTextElement).toHaveTextContent(inputText);
+});
+
+test("should delete an item from todo List", () => {
+  // setup
+  const input = screen.getByPlaceholderText("Please enter todo here");
+  const addButton = screen.getByTestId("add-button");
+  const inputText = "Todo to be deleted!";
+
+  // act
+  userEvent.type(input, inputText);
+  userEvent.click(addButton);
+  const todoList = screen.getByTestId("todo-list");
+  const todoTextElement = getByText(todoList, "Todo to be deleted!");
+  expect(todoList.childElementCount).toBe(2);
+
+  const todoElement = todoTextElement.parentElement;
+  const todoDeleteButton = getByText(todoElement, "Delete");
```

Comment:
as we've added only a single todo item in the _setup_ phase, it's enough to find the Delete button under the `todoList` by its text `Delete`

```js
const todoDeleteButton = getByText(todoList, "Delete");
```

also, prefixing all variables with `todoXXXX` makes the tests more difficult to read, as the context is obvious, it's perfectly fine to just call it `delete`

```js
const delete = getByText(todoList, "Delete");
userEvent.click(delete);
```

Reply:
will remove the setup phase completely as we already have a todo item in our todo list - previous test....if we would add another todo to the list we would not be able to find the respective delete button that easily....

Reply:
`const delete = ' cant be used  as its keyword - will use deleteButton or deleteBtn

Reply:
> will remove the setup phase completely as we already have a todo item in our list - previous test....if we would add another todo to the list we would not be able to find the respective delete button that easily....

Good aiming to remove duplication from test code!

However, what happens if you were to only run this test case without running the first one?

I believe the second test case would fail as the application won't be in the assumed state...

_Test cases should be able to pass independently of each other (ie. when run in separation) and should not depend on a state that was created by previous test cases._

The reason is that it will be very difficult to find out where did something go wrong when test cases start failing. Was it the production code that changed and thus creating the failures? Or was it something in the failing test case? If not which test case changed the application's state in a way that makes other test cases fail?

In these cases a time consuming and daunting debugging will take place stealing time from delivering actual production code / features to customers.

This is the reason why the more test cases we have the more important it becomes for them to be 100% independent.

Suggestion: extract the common "set up" phase to a function which can be called at the beginning of both test cases if you want to remove this duplication. 🙂